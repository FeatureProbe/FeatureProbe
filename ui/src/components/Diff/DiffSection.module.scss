.diff-section {
  padding-top: 16px;

  & > .head-line {
    display: flex;
    flex-grow: 1;

    & > .before,
    & > .after {
      width: 50%;
      margin-left: 24px;

      & > .title {
        display: flex;
        align-items: center;
        font-family: HelveticaNeue-Medium, Arial, Helvetica, sans-serif;
        font-size: 14px;
        color: #212529;
        letter-spacing: 0;
        text-align: justify;
        line-height: 28px;
        font-weight: 500;

        &::before {
          display: block;
          content: ' ';
          width: 3px;
          height: 14px;
          background-color: #556ee6;
          margin-right: 8px;
        }
      }
    }

    & > .before {
      border-right: 1px solid #fff;
    }
  }

  & > .content {
    background-color: #fff;
  }
}
