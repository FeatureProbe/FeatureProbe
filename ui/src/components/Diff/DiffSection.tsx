import { cloneDeep } from 'lodash';
import { ReactNode, useEffect, useMemo, useState } from 'react';
import idiff, { ChangeItem } from './diff';
import styles from './DiffSection.module.scss';

export interface DiffSectionProps {
  title: string;
  before?: unknown;
  after?: unknown;
  diffKey: string;
  renderContent: (diffContent: ChangeItem[]) => ReactNode;
  beforeDiff?: (before: unknown, after: unknown) => unknown[];
  setCount?: (key: string, count: number) => void;
}

const DiffSection: React.FC<DiffSectionProps> = (props) => {
  const { title, before, after, renderContent, beforeDiff, setCount, diffKey } = props;
  const [hide, setHide] = useState(false);

  const diffContent = useMemo(() => {
    let left, right;
    if (beforeDiff) {
      [left, right] = beforeDiff(cloneDeep(before), cloneDeep(after));
    } else {
      [left, right] = [before, after];
    }

    const diffContent = before && after ? idiff(left, right) : undefined;

    return diffContent;
  }, [beforeDiff, before, after]);

  useEffect(() => {
    let modifyCount = 0;
    diffContent?.forEach((item) => {
      if (item.type !== 'same') {
        if (diffKey === 'status' || diffKey === 'default' || diffKey === 'disabled' || diffKey === 'prerequisites') {
          modifyCount = 1;
        } else {
          modifyCount++;
        }
      }
    });

    setHide(modifyCount === 0);
    setCount && setCount(diffKey, modifyCount);
  }, [diffContent, diffKey, setCount]);

  return (
    <div hidden={hide} className={styles['diff-section']}>
      <div className={styles['head-line']}>
        <div className={styles['before']}>
          <div className={styles.title}>{title}</div>
        </div>
        <div className={styles['after']}>
          <div className={styles.title}>{title}</div>
        </div>
      </div>
      <div className={styles.content}>{diffContent && renderContent(diffContent)}</div>
    </div>
  );
};

export default DiffSection;
