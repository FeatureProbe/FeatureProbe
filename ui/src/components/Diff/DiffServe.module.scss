.serve-diff {
  display: flex;
  background-color: #f8f9fa;

  & tr,
  & td,
  & table {
    border: 0 !important;
  }

  & > * {
    width: 100%;
    margin-left: 24px;
    overflow-x: auto;
    background-color: #fff;

    &::-webkit-scrollbar {
      height: 8px;
    }

    & > * {
      display: flex;
      align-items: center;

      & > * {
        margin-right: 12px;
      }
    }
  }
}
