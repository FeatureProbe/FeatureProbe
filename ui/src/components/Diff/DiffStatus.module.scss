.diff-status-content {
  display: flex;
  background: #f8f9fa;
}

.diff-status-item {
  width: 50%;
  padding-left: 24px;
  padding-right: 8px;
  height: 28px;
  border-right: 1px solid #fff;
}

.diff-status-field {
  display: flex;
  align-items: center;
  width: 100%;
  height: 100%;
  background: #fff;
  padding-left: 8px;
}
