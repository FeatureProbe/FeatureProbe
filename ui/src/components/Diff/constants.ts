export type diffType = 'remove' | 'modify' | 'add' | 'same';
export type positionType = 'before' | 'after';

export const rulesI18NMap = new Map([
    ['string', 'targeting.rule.operator.type.string'],
    ['number', 'targeting.rule.operator.type.number'],
    ['datetime', 'targeting.rule.operator.type.datetime'],
    ['semver', 'targeting.rule.operator.type.semver'],
    ['segment', 'targeting.rule.operator.type.segment'],
    ['is one of', 'targeting.rule.condition.isoneof'],
    ['is not any of', 'targeting.rule.condition.notanyof'],
    ['starts with', 'targeting.rule.condition.startswith'],
    ['does not start with', 'targeting.rule.condition.notstartwith'],
    ['ends with', 'targeting.rule.condition.endswidth'],
    ['does not end with', 'targeting.rule.condition.notendwith'],
    ['contains', 'targeting.rule.condition.contains'],
    ['does not contain', 'targeting.rule.condition.notcontain'],
    ['matches regex', 'targeting.rule.condition.matches'],
    ['does not match regex', 'targeting.rule.condition.notmatch'],
    ['before', 'targeting.rule.subject.datetime.before'],
    ['after', 'targeting.rule.subject.datetime.after'],
    ['is in', 'targeting.rule.subject.segment.in'],
    ['is not in', 'targeting.rule.subject.segment.notin'],
    ['between', 'targeting.rule.condition.between']
]);
