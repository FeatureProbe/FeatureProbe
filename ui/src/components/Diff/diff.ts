import { diffArrays, ArrayChange } from 'diff';

const diff = {
  diffArrays,
};

export interface DiffObj {
  [x: string]: unknown;
}

export interface ArrayObj {
  __key?: string;
  __value?: unknown;
}

export type DiffResult = ChangeItem[];

export type ChangeType = 'remove' | 'add' | 'modify' | 'same';

export const displayMap = new Map([
  ['>=', '['],
  ['>', '('],
  ['<', ')'],
  ['<=', ']'],
]);

export interface ChangeItem {
  type: ChangeType;
  value?: unknown;
  diff?: ChangeItem[];
}

//compare in diff function
const compared = (obj1: DiffObj | ArrayObj, obj2: DiffObj | ArrayObj) => {
  if (typeof obj1 === 'object' && typeof obj2 === 'object') {
    if (obj1.__key !== undefined && obj2.__key !== undefined) {
      if (typeof obj1.__value === 'object' && typeof obj2.__value === 'object' && obj1.__value && obj2.__value) {
        return diffObj(obj1.__value as DiffObj, obj2.__value as DiffObj).length === 1;
      }
      return obj1.__value === obj2.__value;
    } else if ((obj1 as DiffObj)['conditions']) {
      return false;
    }
    return diffObj(obj1 as DiffObj, obj2 as DiffObj).length === 1;
  } else {
    return obj1 === obj2;
  }
};

//diff any object or array
function diffObj(left: unknown, right: unknown): ArrayChange<ArrayObj | DiffObj>[] {
  if (left instanceof Array && right instanceof Array) {
    return diff.diffArrays<ArrayObj, ArrayObj>(left as ArrayObj[], right as ArrayObj[], {
      comparator: compared,
    });
  } else {
    const leftArr = Object.keys(left as object)
      .map((key) => {
        if (key === 'objects' && (left as DiffObj)['leftPredicate'] ) {
          return { 
            __key: key, 
            __value: 
              '' + 
              displayMap.get((left as DiffObj)['leftPredicate'] as string) + 
              (left as DiffObj)['objects'] + 
              ', ' + 
              (left as DiffObj)['rightObjects'] + 
              displayMap.get((left as DiffObj)['rightPredicate'] as string)
            };
        } else {
          return { __key: key, __value: (left as DiffObj)[key] };
        }
      })
      .filter((item) => {
        return item.__key !== 'leftPredicate' && item.__key !== 'rightPredicate' && item.__key !== 'rightObjects';
      })
      .sort((a, b) => {
        return a.__key > b.__key ? 1 : -1;
      });
    const rightArr = Object.keys(right as object)
      .map((key) => {
        if (key === 'objects' && (right as DiffObj)['leftPredicate']) {
          return { 
            __key: key, 
            __value: 
              '' + 
              displayMap.get((right as DiffObj)['leftPredicate'] as string) + 
              (right as DiffObj)['objects'] + 
              ', ' + 
              (right as DiffObj)['rightObjects'] + 
              displayMap.get((right as DiffObj)['rightPredicate'] as string)
            };
        } else {
          return { __key: key, __value: (right as DiffObj)[key] };
        }
      })
      .filter((item) => {
        return item.__key !== 'leftPredicate' && item.__key !== 'rightPredicate' && item.__key !== 'rightObjects';
      })
      .sort((a, b) => {
        return a.__key > b.__key ? 1 : -1;
      });

    return diff.diffArrays(leftArr, rightArr, {
      comparator: compared,
    });
  }
}

//Determine whether to continue to diff value
const isTryToDiff = (value: ArrayObj) => {
  if (typeof value === 'object') {
    if (value.__key !== undefined && value.__value !== undefined) {
      if (typeof value.__value === 'object') {
        return {
          flag: true,
          type: 'key-value',
        };
      }
    } else {
      return {
        flag: true,
        type: 'object',
      };
    }
  }
  return {
    flag: false,
  };
};

//reset result to modified
const resetDiffResult = (diffContent: ArrayChange<DiffObj | ArrayObj>[]) => {
  let result: ChangeItem[] = [];

  for (let i = 0; i < diffContent.length; i++) {
    if (diffContent[i].removed) {
      if (i + 1 < diffContent.length && diffContent[i + 1].added) {
        result = [...result, ...toModified(diffContent[i].value as DiffObj[], diffContent[i + 1].value as DiffObj[])];
        i++;
      } else {
        result = result.concat(diffContent[i].value.map((item) => {
          return {
            type: 'remove',
            value: item
          };
        }));
      }
    } else {
      result = result.concat(diffContent[i].value.map((item) => {
        let type: ChangeType = 'same';
        if(diffContent[i].added) {
          type = 'add';
        }
        if(diffContent[i].removed) {
          type = 'remove';
        }
        return {
          type,
          value: item
        };
      }));
    }
  }

  return result;
};

function setRemain(
  removed: DiffObj[],
  added: DiffObj[],
  result: ChangeItem[],
  index: number
) {
  if(index < removed.length) {
    for(let i = index; i < removed.length; i++) {
      result.push({
        type: 'remove',
        value: removed[i]
      });
    }
  }

  if(index < added.length) {
    for(let i = index; i < added.length; i++) {
      result.push({
        type: 'add',
        value: added[i]
      });
    }
  }
}

//let remove + add to modify
function toModified(removed: DiffObj[], added: DiffObj[]) {
  const result: ChangeItem[] = [];
  let i = 0;
  for (i; i < removed.length; i++) {
    if (i >= added.length) {
      break;
    }
    const left = removed[i];
    const right = added[i];

    const isDiff = isTryToDiff(right);
    if (isDiff.flag) {
      if (isDiff.type === 'key-value' && left.__key === right.__key) {
        const diff = diffObj(left.__value, right.__value);
        const type = diff.length === 1 && !diff[0].added && !diff[0].removed ? 'same' : 'modify';

        result.push({
          type,
          value: right,
          diff: type === 'modify' ? resetDiffResult(diff) : undefined
        });
      } else if (left.__key !== right.__key) {
        result.push({
          type: 'remove',
          value: left,
        });
        result.push({
          type: 'add',
          value: right,
        });
      } else {
        const diff = diffObj(left, right);
        const type = diff.length === 1 && !diff[0].added && !diff[0].removed ? 'same' : 'modify';

        result.push({
          type,
          value: right,
          diff: type === 'modify' ? resetDiffResult(diff) : undefined
        });
      }
    } else {
      result.push({
        type: 'remove',
        value: left,
      });
      result.push({
        type: 'add',
        value: right,
      });
    }
  }

  setRemain(removed, added, result, i);

  return result;
}

function idiff(left: unknown, right: unknown): ChangeItem[] {
  return resetDiffResult(diffObj(left, right));
}

export default idiff;
