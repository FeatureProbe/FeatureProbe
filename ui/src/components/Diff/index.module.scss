.diff {
  height: 100%;
  background-color: #f8f9fa;
  padding-bottom: 24px;
}

.warning-circle {
  color: #0a70f5;
  margin-right: 8px;
  font-size: 17px !important;
}

.tips {
  display: flex;
  justify-content: space-between;
  padding: 6px 0 6px 14px;
  font-family: PingFangSC-Regular, Arial, Helvetica, sans-serif;
  font-size: 12px;
  color: #0a70f5;
  line-height: 20px;
  background-color: #eff8ff;
  cursor: pointer;

  & > * {
    display: flex;

    &:last-child {
      & > * {
        display: flex;
        align-items: center;
        margin-right: 16px;
      }
    }
  }
}

.box {
  margin: 0 24px;
  border-radius: 8px;
  overflow: hidden;
}

.old-square {
  width: 12px;
  height: 12px;
  background: #f6d5da;
  border-radius: 4px;
  margin-right: 4px;
}

.new-square {
  width: 12px;
  height: 12px;
  background: #c5f0db;
  border-radius: 4px;
  margin-right: 4px;
}
