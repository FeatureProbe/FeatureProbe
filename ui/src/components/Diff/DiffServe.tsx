import { FormattedMessage } from 'react-intl';
import { CSSProperties } from 'react';
import { Table } from 'semantic-ui-react';
import { ChangeItem } from './diff';
import { DiffFieldValue } from './DiffFieldValue';
import { renderField, renderFieldsItems } from './renderDiff';
import { diffType, positionType } from './constants';
import conditionStyles from './RulesDiffContent.module.scss';
import fieldStyles from './fields.module.scss';
import styles from './DiffServe.module.scss';

interface DiffServeContentProps {
  map?: Map<string, string>;
  content:
    | {
        split?: number[];
        select?: string;
      }
    | ChangeItem[];
  type: positionType;
  diffType: diffType;
  rowStyle?: CSSProperties;
}

export const DiffServeContent: React.FC<DiffServeContentProps> = (props) => {
  const { content, type, diffType } = props;

  const render = () => {
    if (content instanceof Array) {
      return renderField(content, type, (value) => {
        if (value.get('select')) {
          const select = value.get('select');
          if (select?.type === 'add' && type === 'after') {
            return <DiffFieldValue type="add" value={select.value as string} />;
          } else if (select?.type === 'remove' && type === 'before') {
            return <DiffFieldValue type="remove" value={select.value as string} />;
          } else if (select?.type === 'same') {
            return <DiffFieldValue type="same" value={select.value as string} />;
          }
        }
        if (value.get('split')) {
          const split = value.get('split');
          if (split && split.value instanceof Array && typeof split.value[0] !== 'object') {
            return <DiffFieldValue type={split.type} value={split.value} />;
          } else {
            return renderFieldsItems(split?.value as ChangeItem[], type, (value, diffType, _, index) => {
              if (diffType === 'add' && type === 'after') {
                return <DiffFieldValue type="add" value={value as string} key={index} />;
              } else if (diffType === 'remove' && type === 'before') {
                return <DiffFieldValue type="remove" value={value as string} key={index} />;
              } else if (diffType === 'same') {
                return (
                  <DiffFieldValue
                    style={{
                      background: 'rgba(33,37,41,0.08)',
                    }}
                    type="same"
                    value={value as string}
                    key={index}
                  />
                );
              }
            });
          }
        }
      });
    } else {
      if (content.select !== undefined) {
        return <DiffFieldValue value={content.select} />;
      }
      if (content.split !== undefined) {
        return <DiffFieldValue value={content.split as unknown as string[]} />;
      }
    }
  };

  return (
    <Table.Row className={`${fieldStyles[`diff-item-${diffType}`]} ${conditionStyles['condition-diff-item']}`}>
      <Table.Cell>
        <FormattedMessage id="common.serve.text" />
      </Table.Cell>
      <Table.Cell colSpan="5">{render()}</Table.Cell>
    </Table.Row>
  );
};

interface DiffServeProps {
  content:
    | {
        split?: number[];
        select?: string;
      }
    | ChangeItem[];
}

export const DiffServe: React.FC<DiffServeProps> = (props) => {
  const { content } = props;

  return (
    <div className={styles['serve-diff']}>
      <div>
        <Table>
          <Table.Body>
            <DiffServeContent diffType="modify" content={content} type={'before'} />
          </Table.Body>
        </Table>
      </div>
      <div>
        <Table>
          <Table.Body>
            <DiffServeContent diffType="modify" content={content} type={'after'} />
          </Table.Body>
        </Table>
      </div>
    </div>
  );
};
