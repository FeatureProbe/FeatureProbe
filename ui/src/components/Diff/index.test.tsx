import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import Diff from './';
import { IntlWrapper } from 'components/utils/wrapper';
import { DiffStatusContent } from './DiffStatus';
import { useIntl } from 'react-intl';

const Component = () => {
  const intl = useIntl();
  return (
    <Diff
      sections={[
        {
          before: {
            disabled: true,
          },
          after: {
            disabled: false,
          },
          title: intl.formatMessage({ id: 'targeting.status.text' }),
          renderContent: (content) => {
            return <DiffStatusContent content={content} />;
          },
          diffKey: 'status',
        },
      ]}
    />
  );
};

it('Diff snapshot', (done) => {
  (async () => {
    const { asFragment } = render(<Component />, {
      wrapper: IntlWrapper,
    });
    expect(asFragment()).toMatchSnapshot();

    const tips = screen.getByText('The following is a before and after comparison of this change.');
    await userEvent.click(tips);
    expect(asFragment()).toMatchSnapshot();

    done();
  })();
});
