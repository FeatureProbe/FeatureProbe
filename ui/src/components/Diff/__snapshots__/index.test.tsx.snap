// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Diff snapshot 1`] = `
<DocumentFragment>
  <div
    class="box"
  >
    <div
      class="tips"
    >
      <div>
        <span
          class="iconfont icon-warning-circle warning-circle"
          customclass="warning-circle"
          type="warning-circle"
        />
        The following is a before and after comparison of this change.
        <span>
          (1 changes)
        </span>
      </div>
      <div>
        <div>
          <div
            class="old-square"
          />
          Old/Delete
        </div>
        <div>
          <div
            class="new-square"
          />
          New/Create
        </div>
        <div>
          <span
            class="iconfont icon-angle-down icon-accordion"
            customclass="icon-accordion"
            type="angle-down"
          />
        </div>
      </div>
    </div>
    <div
      class="diff"
      style="display: none;"
    >
      <div
        class="diff-section"
      >
        <div
          class="head-line"
        >
          <div
            class="before"
          >
            <div
              class="title"
            >
              Status
            </div>
          </div>
          <div
            class="after"
          >
            <div
              class="title"
            >
              Status
            </div>
          </div>
        </div>
        <div
          class="content"
        >
          <div
            class="diff-status-content"
          >
            <div
              class="diff-status-item"
            >
              <div
                class="diff-status-field"
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  disabled
                </div>
              </div>
            </div>
            <div
              class="diff-status-item"
            >
              <div
                class="diff-status-field"
              >
                <div
                  class="diff-field-value field-value-add"
                >
                  enabled
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Diff snapshot 2`] = `
<DocumentFragment>
  <div
    class="box"
  >
    <div
      class="tips"
    >
      <div>
        <span
          class="iconfont icon-warning-circle warning-circle"
          customclass="warning-circle"
          type="warning-circle"
        />
        The following is a before and after comparison of this change.
        <span>
          (1 changes)
        </span>
      </div>
      <div>
        <div>
          <div
            class="old-square"
          />
          Old/Delete
        </div>
        <div>
          <div
            class="new-square"
          />
          New/Create
        </div>
        <div>
          <span
            class="iconfont icon-angle-up icon-accordion"
            customclass="icon-accordion"
            type="angle-up"
          />
        </div>
      </div>
    </div>
    <div
      class="diff"
      style="display: block;"
    >
      <div
        class="diff-section"
      >
        <div
          class="head-line"
        >
          <div
            class="before"
          >
            <div
              class="title"
            >
              Status
            </div>
          </div>
          <div
            class="after"
          >
            <div
              class="title"
            >
              Status
            </div>
          </div>
        </div>
        <div
          class="content"
        >
          <div
            class="diff-status-content"
          >
            <div
              class="diff-status-item"
            >
              <div
                class="diff-status-field"
              >
                <div
                  class="diff-field-value field-value-remove"
                >
                  disabled
                </div>
              </div>
            </div>
            <div
              class="diff-status-item"
            >
              <div
                class="diff-status-field"
              >
                <div
                  class="diff-field-value field-value-add"
                >
                  enabled
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
